mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -226,55 +225,95 @@ impl State for WaitForTakerPaymentSpend {
-        if std::env::var("SWAP_WATCHER_SKIP_WAITING").is_err() {
-            // Sleep for half the locktime to allow the taker to spend the maker payment first
-            let now = now_ms() / 1000;
-            let wait_for_taker_until =
-                wait_for_taker_payment_conf_until(watcher_ctx.data.swap_started_at, watcher_ctx.data.lock_duration);
-            let sleep_duration = (wait_for_taker_until - now + 1) as f64;
-
-            if now < wait_for_taker_until {
-                Timer::sleep(sleep_duration).await;
-            }
-        }
-
-        let f = watcher_ctx.taker_coin.wait_for_htlc_tx_spend(
-            &watcher_ctx.data.taker_payment_hex,
-            &[],
-            watcher_ctx.data.taker_payment_lock,
-            watcher_ctx.data.taker_coin_start_block,
-            &None,
-        );
-
-        let tx = match f.compat().await {
-            Ok(t) => t,
-            Err(err) => {
-                error!("{}", err.get_plain_text_format());
-                return Self::change_state(RefundTakerPayment {});
-            },
+        const TAKER_PAYMENT_SPEND_SEARCH_INTERVAL_SLOW: f64 = 300.;
+        let payment_search_interval = match std::env::var("USE_TEST_SEARCH_INTERVAL") {
+            Ok(_) => TAKER_PAYMENT_SPEND_SEARCH_INTERVAL,
+            Err(_) => TAKER_PAYMENT_SPEND_SEARCH_INTERVAL_SLOW,
         };
-
-        let tx_hash = tx.tx_hash();
-        info!("Taker payment spend tx {:02x}", tx_hash);
-        let tx_ident = TransactionIdentifier {
-            tx_hex: BytesJson::from(tx.tx_hex()),
-            tx_hash,
+        let wait_until = match std::env::var("SKIP_WAIT_FOR_TAKER_PAYMENT") {
+            Ok(_) => 0,
+            Err(_) => watcher_ctx.data.swap_started_at + (4 * watcher_ctx.data.lock_duration / 3),
+        };
+        let search_input = WatcherSearchForSwapTxSpendInput {
+            time_lock: (watcher_ctx.data.swap_started_at + watcher_ctx.data.lock_duration) as u32,
+            taker_pub: &watcher_ctx.verified_pub,
+            maker_pub: &watcher_ctx.data.maker_pub,
+            secret_hash: &watcher_ctx.data.secret_hash,
+            tx: &self.taker_payment_hex,
+            search_from_block: watcher_ctx.data.taker_coin_start_block,
+            wait_until,
+            check_every: payment_search_interval,
         };
 
-        let secret = match watcher_ctx
+        let f = watcher_ctx
             .taker_coin
-            .extract_secret(&watcher_ctx.data.secret_hash, &tx_ident.tx_hex.0)
-            .await
-        {
-            Ok(bytes) => H256Json::from(bytes.as_slice()),
+            .watcher_search_for_swap_tx_spend(search_input)
+            .await;
+
+        match f {
+            Ok(FoundSwapTxSpend::Spent(tx)) => {
+                let now = now_ms() / 1000;
+                if now < watcher_ctx.data.swap_started_at + watcher_ctx.data.lock_duration {
```

Comment:
Is this really required? I suppose that if `now >= watcher_ctx.data.swap_started_at + watcher_ctx.data.lock_duration` then `wait_for_htlc_tx_spend` will return an error. Removing this condition will reduce nesting a bit.

Reply:
If the taker payment is spent by the maker before `watcher_ctx.data.swap_started_at + watcher_ctx.data.lock_duration`, the watcher retrieves the maker payment hex by using the `get_tx_hex_by_hash`, then goes to the `wait_for_htlc_tx_spend`. Having the extra check here reduces a `get_tx_hex_by_hash` request which will be skipped if the condition does not hold.